### PR TITLE
Type hints for lib module

### PR DESCRIPTION
**Pull request by umak1106**
Fixes #

Changes made in this Pull Request:
 - 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by umak1106** [comment]
Regarding the stp() function if i set the arguments as arraylike i get this error
Argument 1 to "cross" has incompatible type "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"; expected "Union[_SupportsArray[dtype[bool_]], _NestedSequence[_SupportsArray[dtype[bool_]]], bool, _NestedSequence[bool]]"

And if i set it as NDArray I still get an error
Argument 1 to "stp" has incompatible type "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"; expected "ndarray[Any, dtype[Any]]"

The same error above shows  for normal()

**Comment by umak1106** [comment]
actually typing module is introduced in numpy from 1.20.0 and the gh-ci.yaml file has numpy = 1.19.0 which is why i think the tests are failing

**Comment by jbarnoud** [comment]
@IAlibay plans on bumping the minimal version of numpy to 1.20.0. It means we'll have to wait for the bump to happen before we merge, but you can proceed and ignore the failing pipeline.

File: package/MDAnalysis/lib/mdamath.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/mdamath.py]: Here, the inputs can be `ArrayLike`, I think.

The idea is to be loose on the input to accommodate everything that is valid, but precise on the output.

File: package/MDAnalysis/lib/mdamath.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/mdamath.py]: Here too, I think the input can be `ArrayLike`. It should be true for other functions as well.

**Review by jbarnoud — COMMENTED** [review]
Do you still have the error you described with `normal` and `stp`? I did not get them when testing locally.

I only have small comments about the typing itself. In general, we want to be broad when it comes to inputs and as specific as possible for the outputs. This is because we do not know what we will get as inputs and will accept anything that can work, but (most of the time) we know exactly what we will output.

You should remove the files you annotate from those ignored by mypy so that the CI actually runs mypy on the files you edit.

File: package/MDAnalysis/lib/mdamath.py
Comment (umak1106) [reply package/MDAnalysis/lib/mdamath.py]: Ok I'll make the changes

File: package/MDAnalysis/lib/mdamath.py
Comment (umak1106) [reply package/MDAnalysis/lib/mdamath.py]: Actually the doctoring said ndarray so I did that but I'll make the changes.
After making the changes I get error on that line

Comment (jbarnoud) [reply package/MDAnalysis/lib/mdamath.py]: Automatic tools may tell you that, indeed, because it is how it is use in the examples it seen. But if you read the code, you see that `a` and `b`  are passed to the `np.einsum`function that expects an `ArrayLike` there.

**Comment by hmacdope** [comment]
@umak1106 just a quick note, I am adding type hints to a lot of `distances.py` as part of #3730, you can probably leave that one to me for now.

**Comment by umak1106** [comment]
ok

File: package/MDAnalysis/__init__.py
Comment (jbarnoud) [inline package/MDAnalysis/__init__.py]: Is there a reason for the different style of annotation? If not, I'd rather have the style consistent:

```python
_READERS: Dict = {}
```

Also, the annotation can be more specific:

```python
_READERS: Dict[str, ProtoReader] = {}
```

Comment (umak1106) [reply package/MDAnalysis/__init__.py]: Ok I'll make the changes

Comment (umak1106) [reply package/MDAnalysis/__init__.py]: I made the changes like you suggested . But the tests are failing

Comment (jbarnoud) [reply package/MDAnalysis/__init__.py]: I looked at it and it looks like the issue is caused by the imports in `__init__.py`.

I tried two ways of fixing the problem:
* I guarded the import with `TYPE_CHECKING`, but then I needed to reference the type with forward references (using the name of the types as strings instead of just providing the types)
* providing the types in a stub file (see <https://mypy.readthedocs.io/en/stable/stubs.html>)

Both methods seem to work in the sense that nowhere mypy complains about not having annotations for these variables. However, I know the annotations are wrong, yet I do not get any errors about them. So I would suggest keeping it simple for now and just annotating them as `Dict`. It will not be as precise as it should, but at least it won't be wrong.

**Comment by umak1106** [comment]
Errors am facing :
MDAnalysis/lib/pkdtree.py:141: error: Incompatible types in assignment (expression has type "Optional[float]", variable has type "None")
MDAnalysis/lib/pkdtree.py:184: error: Unsupported operand types for > ("float" and "None")
MDAnalysis/lib/pkdtree.py:234: error: Unsupported operand types for > ("float" and "None")
MDAnalysis/lib/pkdtree.py:290: error: Unsupported operand types for > ("float" and "None")

File: package/MDAnalysis/lib/pkdtree.py
Comment (umak1106) [inline package/MDAnalysis/lib/pkdtree.py]: according to the docstring self._indices should be a list but doesn't np.asarray return NDArray

File: package/MDAnalysis/lib/pkdtree.py
Comment (umak1106) [inline package/MDAnalysis/lib/pkdtree.py]: Same as above

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: Looks like the docstring is out of date, indeed. Could you fix it?

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: Here the docstring and the code seem to match, don't they? `pair` is an array is I recall correctly.

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]:
```suggestion
    def set_coords(self, coords: npt.ArrayLike, cutoff: Optional[float] = None) -> None:
```

**Review by jbarnoud — COMMENTED** [review]
Regarding the errors you report:
* for the first one, you probably should type the instance attribute. See https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes
* for the other errors, we assume that cutoff is set if pbc is True. It holds with the logic of how we expect the class to be used, but it can break, which is what mypy is warning us about. You may need to add a few checks, here and there, to make sure that cutoff is not None when pbc is True.

File: package/MDAnalysis/lib/pkdtree.py
Comment (umak1106) [reply package/MDAnalysis/lib/pkdtree.py]: Yes I misread . The doctoring mentions array .

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]: `self.cutoff` can be `None` is `self.pbc` is `False`. It is only if `self.pbc == True` that we assume we have a value for the cutoff.

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]: I *think* a `ValueError` would be more appropriate here.

**Review by jbarnoud — CHANGES_REQUESTED** [review]
Last few small things:

* there is a small misunderstanding with the test about cutoff. I explained this in a comment at one of the places you added the test but it is true for all the places where you have that test.
* there seems to be an issue with using `npt.NDArray` and numpy 1.20.0. See #3748. For the time being, can you use `np.ndarray` instead so we can move forward?

File: package/MDAnalysis/lib/pkdtree.py
Comment (umak1106) [reply package/MDAnalysis/lib/pkdtree.py]: If i use 
if self.pbc and (self.cutoff is None) :
raise Valuerror("Provide cutoff")

mypy still raises this error 
Unsupported operand types for > ("float" and "None")
Left operand is of type "Optional[float]"

Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: Can you push the code with the conditional moved? This way I can comment on the code you have rather than the code I imagine you may have. Also, it will let us see if the tests, other than mypy, pass.

**Comment by jbarnoud** [comment]
It looks like you are still using `npt.NDArray` in pkdtree.

**Comment by umak1106** [comment]
> It looks like you are still using `npt.NDArray` in pkdtree.

Yes actually after i faced mypy errors there i didn't push my code

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: This remains outstanding. It looks like it is the latest obstacle in terms of type annotations.

Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: I did not realise you did push your latest change. Here is a possible solution. I did not test it, but it might work:

```python
if self.pbc:
    if self.cutoff is None:
        raise ValueError("Cutoff needs to be provided when working with PBC.")
    cutoff: float = self.cutoff
    # From here and until the end if the if self.pbc block, use cutoff instead of self.cutoff
```

I haven't tested that solution, but I expect that mypy will figure out that `cutoff` can only be a float.

Comment (umak1106) [reply package/MDAnalysis/lib/pkdtree.py]: I tried that but it still raises the same error

Comment (jbarnoud) [reply package/MDAnalysis/lib/pkdtree.py]: I am confused. I tried on a small example and it works perfectly fine:

```python
from typing import Optional

def test(pbc: bool, cutoff: Optional[float]):
    if pbc:
        #if cutoff is None:
        #    raise ValueError
        if cutoff < 3:
            print('Hello')
    print('Good bye')

test(False, 2)  # OK, cutoff is just ignored
test(False, None)  # OK
test(True, 2)  # OK
test(True, None)  # Not OK
```

With the `if cutoff is None` commented, I get the same mypy error:

```
test.py:7: error: Unsupported operand types for > ("int" and "None")
test.py:7: note: Left operand is of type "Optional[float]"
```

If it is uncommented, I do not get any errors.

Comment (umak1106) [reply package/MDAnalysis/lib/pkdtree.py]: Done

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]: Here, the original test is not about `self.cutoff`, it is instead about `cutoff` as provided as an argument to the function. This seems to be why the tests fail.

```suggestion
            if cutoff is None:
```

It is subtle and got me puzzled for a while!

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]: Move the comment with the code. Now we have an orphan comment that will confuse future contributors and maintainers.

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py]: In this specific instance, because you move an existing exception raising, it would be good to raise the same exception as what you removed:

```suggestion
                raise RuntimeError('Provide a cutoff distance'
                               ' with tree.set_coords(...)'
```

File: package/MDAnalysis/lib/pkdtree.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/pkdtree.py:139]: This line is extremely important and cannot be removed! This line actually sets the cutoff in the object; without it, `self.cutoff` will always be `None`. The line needs to be run if the exception about the cutoff is not raised.

File: package/MDAnalysis/lib/NeighborSearch.py
Comment (jbarnoud) [inline package/MDAnalysis/lib/NeighborSearch.py]:
```suggestion
    def _index2level(self, indices: List[int], level: str) -> Union[AtomGroup, ResidueGroup, SegmentGroup]:
```

**Review by jbarnoud — CHANGES_REQUESTED** [review]
I made some suggestions to fix errors that caused either the tests to fails or mypy to complain. After that, all the complaints from mypy are about `distances.py` that have wrong annotations. I'll try and fix them.

**Comment by jbarnoud** [comment]
Now that #3787 is merged into develop, the develop branch should be merged in here.

**Comment by jbarnoud** [comment]
Wooo! 🎉 Only one mypy error left! With the last few comment addressed we should be good to go 😁

**Comment by umak1106** [comment]
Making those changes in pkdtree raises errors like unresolved reference to self

**Comment by jbarnoud** [comment]
> Making those changes in pkdtree raises errors like unresolved reference to self

Can you paste the full error message please? It is difficult to figure out what can go wrong without it.

File: package/MDAnalysis/__init__.py
Comment (jbarnoud) [inline package/MDAnalysis/__init__.py]: @umak1106 It would be great to have these lines in a separate PR so we can work at annotating other parts of the code without triggering errors here.

Comment (umak1106) [reply package/MDAnalysis/__init__.py]: ok mentor.
 #3823 this is the pr .

File: package/MDAnalysis/lib/pkdtree.py
Comment (umak1106) [reply package/MDAnalysis/lib/pkdtree.py]: Done

**Comment by umak1106** [comment]
> > Making those changes in pkdtree raises errors like unresolved reference to self
> 
> Can you paste the full error message please? It is difficult to figure out what can go wrong without it.

It got fixed . There are no mypy errors in this module now .

**Review by jbarnoud — APPROVED** [review]
Youhou! :tada:

**Comment by jbarnoud** [comment]
Congratulations on that PR being merged!

**Comment by umak1106** [comment]
> Congratulations on that PR being merged!

Thanks

**Pull request merged**
